File: java/Day01.java
import java.util.Scanner;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.io.File;

public class Day01 {

    public static int getDifference(String file) {
        ArrayList<Integer> left = new ArrayList<>();
        ArrayList<Integer> right = new ArrayList<>();
        // organize into a left number and right number list
        try (Scanner sc = new Scanner(new File(file))){
            while (sc.hasNextLine()) {
                String[] line = sc.nextLine().split("\\s+");
                left.add(Integer.parseInt(line[0]));
                right.add(Integer.parseInt(line[1]));
            }
        } catch (FileNotFoundException e) {
            System.out.println("error, no file");
        }
        // sort the inputs, use insertion sort
        Collections.sort(left);
        Collections.sort(right);
        // get the frequency array of the right
        // index (of number), freq (of number)
        Map<Integer, Integer> freq = new HashMap<>();
        for (Integer i : right){
            if (!freq.containsKey(i)) {
                freq.put(i, 0);
            }
            freq.put(i, freq.get(i)+1);
        }
        int sim = 0;
        for (Integer i : left) {
            if (freq.containsKey(i)) {
                sim += i * freq.get(i);
            }
            
        }
        System.out.println("Sim: " + sim);

        // get difference of each and add it to the sum
        int sum = 0;
        for (int i = 0; i < left.size(); i++) {
            sum += Math.abs(left.get(i) - right.get(i));
        }
        System.out.println("Sum: " + sum);
        return sum;
    }
    


    public static void main(String[] args) {
        System.out.println(getDifference("C:\\Users\\abby-\\projects\\AdventOfCode\\days\\input.txt"));
    }
}

File: js/day01.js
const fs = require('fs');

function getAnswer() {
    const left = [];
    const right = [];
    const freq = {};
    // read the file in
    const lines = fs.readFileSync('../input.txt').toString().split('\n');
    // split the input up into a left and right list
    lines.forEach((s) => {
        const[l, r] = s.split(/\s+/).map(s => parseInt(s));

        left.push(l);
        right.push(r);
        freq[r] = (freq[r] ?? 0) + 1;
    });
    // sort the lists
    left.sort();
    right.sort();
    // get the total difference of the lists
    diffSum = 0;
    for (let i = 0; i < left.length; i++) {
        diffSum += Math.abs(left[i] - right[i]);
    }
    console.log("difference sum: " + diffSum);

    // get the similarity score
    let similarity = 0;

    left.forEach((num) => {
        similarity += num * (freq[num] ?? 0);
    })
    console.log("similarity: " + similarity);
}

function main() {
    getAnswer();
}

main();